### DISCUSSION TITLE
How to realize the element average of a MaterialProperty in the material class?

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#19648] [Q&A General] [answered]
Dear MOOSE experts,

The `MaterialProperty` average value of a certain calculation domain is needed as another material parameter. So, I am trying to find the element average or node average of a `MaterialProperty ` in the material module. I focused on referring to the `ElementAverageMaterialProperty` class in the Postprocessors module. 

`ElementAverageMaterialPropertyTest` was created to implement such operations, the code is as follows,
```c++
// ElementAverageMaterialPropertyTest.h
protected:
MaterialProperty<Real> & _material_property;

void computeIntegral(Real & material_property); // Real & _material_property

template <typename T>
void gatherSum(T & value)
{
  _communicator.sum(value);
}
```

```c++
// ElementAverageMaterialPropertyTest.C
ElementAverageMaterialPropertyTest::ElementAverageMaterialPropertyTest(const InputParameters & parameters):_material_property(declareProperty<Real>(_base_name + "material_property"))
{    
}

void
ElementAverageMaterialPropertyTest::computeIntegral(Real & material_property)
{
  std::vector<Real> sum(2, 0);
  std::vector<Real> area(2, 0); 
  Real average = 0.0;

  unsigned int op_index = 0;
  unsigned int size = _hard_factor.size();

    //  Calculate the average for each element
  for (unsigned int qp = 0; qp < _qrule->n_points(); qp++) // _qrule->n_points()--四节点单元网格
  {
    sum[op_index] += _JxW[qp] * _coord[qp] * material_property; // Calculate the average of each element
    area[op_index] += _JxW[qp] * _coord[qp]; // Calculate the area of each element

    // _JxW[qp] = 0.03; // the area of each element is 0.03*4 = 0.12 for QUAD4
    // _coord[qp] = 1;
    // _qrule->n_points() = 4;
  }

    _communicator.sum(sum[op_index]);
    _communicator.sum(area[op_index]);

    average = sum[0]/area[0];
}
```
When using 30 Processors, an error occurred during the calculation process, the error is as follows,
```c++
application called MPI_Abort(MPI_COMM_WORLD, 805906695) - process 9
Generic Warning: In /opt/civet/build_1/conda_builds/conda_envs/next-2e4c730e2f/conda-bld/moose-libmesh-vtk_1626837403905/work/Parallel/MPI/vtkMPICommunicator.cxx, line 69
MPI had an error
------------------------------------------------
Invalid root, error stack:
PMPI_Bcast(448): MPI_Bcast(buf=0x7ffe66f8efa0, count=1, MPI_UNSIGNED_LONG, root=1074329026, comm=MPI_COMM_WORLD) failed
PMPI_Bcast(408): Invalid root (value given was 1074329026)
------------------------------------------------

application called MPI_Abort(MPI_COMM_WORLD, 269035783) - process 22
```

Finally, It should be noted that if I cancel the following code, it can run normally.
```c++
    _communicator.sum(sum[op_index]);
    _communicator.sum(area[op_index]);
```
I think I don't understand the execution mechanism of this code very well. In other words, I don't understand how to traverse all the finite element mesh operations in the material class.

Any suggestions or recommendations to fix the problem would be greatly appreciated.

Thank you
Wei

**Comment by dschwen** [comment, marked as answer]
You must _not_ perform MPI communication in the `execute()` method of an element userobject/postprocessor! `computeIntegral` is a virtual function that gets called by `execute`. `execute` is run on each element, and there is very likely not the same number of local elements on each processor, which means there are bound to be MPI calls that have no corresponding call on all other processors, which leads to failure.

I'm not entirely clear what you want to do here. Do you need the average over elements or the average over a (sub)domain? The average over the element can be computed in `computeProperties` by looping over all QPs an additional time. If you need the average over the domain (and I'm curious what physics would require such a non-local effect) you need to couple in a postprocessor value, and let MOOSE compute the average using `ElementAverageMaterialProperty`.

**Reply by PengWei97** [reply]
Thank you for your kind help. 

> You must not perform MPI communication, ..., which leads to failure.

In paragraph, `ElementAverageMaterialPropertyTempl<is_ad>::execute()` is responsible for calculating the sum value on each process or thread on a subset of the domain by calling the `MaterialProperty` on each point. At this stage, MPI communication is not performed, right? 

After that, the actual summations operation is done in `ElementAverageMaterialPropertyTempl<is_ad>::getValue()` througth the function `gatherSum()`. MPI needs to be called. 

I don't know if my understanding is correct?

Finally, I dinn't see the relevent code for calling between `execute()` and `getvalue()`, so I am not very clear about the calling relationship between the them. And these two related codes are also listed at the back

# execute()
1. ElementAverageMaterialPropertyTempl
```c++
template <bool is_ad> // ElementAverageMaterialPropertyTempl.C
void
ElementAverageMaterialPropertyTempl<is_ad>::execute()
{
  ElementIntegralMaterialPropertyTempl<is_ad>::execute();
  // _integral_value += computeIntegral(); 积分所计算的值

  _volume += this->_current_elem_volume; // 计算所的体积
}
```

2. ElementIntegralPostprocessor
```c++
virtual void execute() override; //*.h
virtual Real computeIntegral();

void
ElementIntegralPostprocessor::execute()
{
  _integral_value += computeIntegral();
}

Real
ElementIntegralPostprocessor::computeIntegral()
{
  Real sum = 0;

  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
    sum += _JxW[_qp] * _coord[_qp] * computeQpIntegral();
  return sum;
}
```

3. ElementIntegralMaterialPropertyTempl
```c++
template <bool is_ad>
Real
ElementIntegralMaterialPropertyTempl<is_ad>::computeQpIntegral()
{
  return MetaPhysicL::raw_value(_scalar[_qp]);
}
```

# getValue()
1. ElementAverageMaterialPropertyTempl
```c++
template <bool is_ad>
Real
ElementAverageMaterialPropertyTempl<is_ad>::getValue()
{
  const Real integral = ElementIntegralMaterialPropertyTempl<is_ad>::getValue();
  // gatherSum(_integral_value); // in ElementIntegralPostprocessor.C

  ElementIntegralMaterialPropertyTempl<is_ad>::gatherSum(_volume);
  // gatherSum:: a gather operation is required to collect the values computed on all processes to the root process;
  
  return integral / _volume;
}
```

2. ElementIntegralPostprocessor
```c++
Real
ElementIntegralPostprocessor::getValue()
{
  gatherSum(_integral_value);
  return _integral_value;
}
```

**Reply by hugary1995** [reply]
Hi Wei,

Could you tell us what you want to compute here? I'll repeat @dschwen here: 

> Do you need the average over elements or the average over a (sub)domain?

Additionally, how do you want to use that average? Do you want to use that average in a material property calculation, in a kernel/auxkernel, or in a postprocessor, etc.? I have a feeling that you are probably going down the wrong path -- There could be a much simpler alternative to what you doing right now.

> Finally, I dinn't see the relevent code for calling between execute() and getvalue(), so I am not very clear about the calling relationship between the them. And these two related codes are also listed at the back

To understand the "calling relationship" you'll have to dig through the parent classes of the postprocessor. There are multiple ways of doing that. We do have a [C++ fundamentals](https://mooseframework.inl.gov/help/c++/index.html) page to help you get started.

**Reply by PengWei97** [reply]
>  I'm curious what physics would require such a non-local effect

I am trying to establish a coupled model of phase field (the diffuse interface description) and tensor mechanics (the shrap interface description), such as J2 plasticity or crystal plasticity, based on moose to study the influence of deformed energy including elastic energy and plastic energy on microstructure evolution, such grain boundary migration or precipitation. Plase forgive me for the slow progress now, considering that I don't understand some of mechanisms in moose.

In the process of coupling modeling, in view of the incompatibility of the two calculation methods in the description of the microstructure, some assumptions need to be introduced at the interface. [L. Zhao, P Chakraborty, M.R. Tonks, I. Szlufarska (2017)](https://www.sciencedirect.com/science/article/pii/S092702561630605X) used an interpolation scheme to achieve the coupling between them. 

In this article, 
![image](https://user-images.githubusercontent.com/56183775/146480555-f110d6a2-f172-44bf-af27-b7fb3dab771d.png)


The sentence marked in green indicates that a weighted average operation is required for the stress and internal states in each grain. The operation of calculating the weighted average of `MaterialProperty` is my original intention to ask this question. This weighted value needs to be transmitted to the materials bolk again for the calculation of kenels.

Finally, this model is also based on MOOSE:MARMOT, but I don't have the permission to read it. So I can only directly to the instructions in the article and try to reproduce its operation in my model.

I truly appreciate your timely help.
Wei

**Reply by hugary1995** [reply]
Okay, this is much clearer. At the grain interface, you will have more than one nonzero order parameter. Since each order parameter represents a crystal (and each crystal has its own stress), you'll have to somehow "blend" them together.

This operation is performed at each quadrature point -- you don't need any additional loop for that. Suppose you have a set of order parameters `phi_i`, then one way to perform the weighted average is
```
stress = sum(weight(phi_i) * stress_i) / sum(weight(phi_i))
```
where `weight(phi_i)` is the weighting function of your choice. I think a common choice is `phi_i^2`.

**Reply by PengWei97** [reply]
Hi @hugary1995,

> I have a feeling that you are probably going down the wrong path -- There could be a much simpler alternative to what you doing right now.

Considering the clarity of the question, I divided the response into two. In the last reply, I gave the reason why I want to perform the averaging of the material parameters within the grain (in a certain domain). [L. Q. Chen (2015)](https://www.sciencedirect.com/science/article/pii/S0045782514004861?casa_token=NytlEjCPzi4AAAAA:vZaepmbFmWHWtYa2pzshsI6hhKH0AzlhynLO0QOYUeYoLsXaQ-A-F-ZcuSGesbefm-6ylHjY4KTL) gives a detailed operation, and I sincerely hope that you have read related articles. 

> To understand the "calling relationship" you'll have to dig through the parent classes of the postprocessor.

Yeah, I also started my introduction to C++ based on [the C++ fundamental](https://mooseframework.inl.gov/help/c++/index.html). But my question now is what is the calling relationship between `execute()` and `getValue()`. In moose, I did not find any relevant lines of code to illustrate the calling relationship between them. `ElementUserObject.C` and `Postprocessor.C`also have no related functions to coordinate the calling relationship between them.

I don’t know if my question is clear, if not, please explain directly.
Wei

**Reply by PengWei97** [reply]
Hi, @hugary1995

Thank you for your kind answer.

> stress = sum(weight(phi_i) * stress_i) / sum(weight(phi_i))；

I absolutely understand what you mean, but the most important question now is how to get `stress_i`. So I want to calculate `stress_i` by weighted average, the `tensor mechanics` block can only get the only certain `stress[ _qp]` at the orthogonal point.

**Reply by hugary1995** [reply]
You can use `MultiPhaseStressMaterial` to do that.

**Reply by hugary1995** [reply]
Well, I guess I'm still not quite sure about what you really want to do, since what you described is different from what was suggested in the article screenshot you pasted above. 

Anyways, I also replied to your other post #19650, in case an average of material property over a grain is actually what you want. The grain average makes sense to me if you want to couple FFT and FEM.

**Reply by PengWei97** [reply]
# problem 1
>  I guess I'm still not quite sure about what you really want to do,…

I give two articles that may cause some confusion.  I mainly refer to zhao's article, and Chen LongQing’s is just as a supplement, in terms of interpolation processing at the interface.

In [L. Zhao, P Chakraborty, M.R. Tonks, I. Szlufarska (2017)](https://www.sciencedirect.com/science/article/pii/S092702561630605X), the processing method at the interface is done by interpolating `stress` and `interal states variables`,
![image](https://user-images.githubusercontent.com/56183775/146539747-2a407975-b12b-4762-bc25-e2669b4575f6.png)
According to my understanding, `a weighted sum of stresses in different crystals` in the words marked in green indicate that a weighted average of the stress in each grain needs to be processed.

A more detailed interpolation scheme is as follows,
![image](https://user-images.githubusercontent.com/56183775/146539667-46bcc3c2-80db-4bc1-a74b-3446f9bb0f6f.png)

Zhao applied the same interpolation method to the Cauchy stress $\sigma_i$, the slip system resistence $s^\alpha_i$ and the deformed energy $\psi^{e}_i,\ \psi^{p}_i$

However, zhao did not give calculation or description details of the ${\sigma}_{i}, s_{i}^{\alpha}, \psi_{i}^{e}, \psi_{i}^{p}}$ corresponding to each crystal or grain, just say a sentence `a weighted sum of stresses in different crystals, ...` . Therefore, I focused on the interpolation scheme mentioned in Long qingchen (2015), Since he explained their interpolation scheme in detail. The following is
![image](https://user-images.githubusercontent.com/56183775/146540216-14c6335b-ebba-410e-b74b-40dade76c2d4.png)

The interpolation method mentioned in the chen Longqing article is just an alternative for me, because I am not sure how to calculate the stress and internal parameters of each grain in zhao. According to my understanding, the stress and internal parameters calculated by the tensor mechanics module are uniquely determined at each orthogonal point, without distinguishing which crystal grain they belong to.

And today you said that I can try to use `MultiPhaseStressMaterial` to realize the stress interpolation process at the interface, and I carefully looked at two examples involving `MultiPhaseStressMaterial`, `multiphasestress.i` and `Pattern1.i`. 

I guess Zhao used this coupling modeling method to a large extent. In detail, the stress of each grain is calculated in the material module, and then they are weighted and averaged with `MultiPhaseStressMaterial`, and finally transmitted to to calculate the balance equation of the mechanics field in kernels. Moreover, I think the advantage of doing so is that the mechanical constitutive model used can be modified very flexibly. 

However, I did not see how this method can reset the state parameters at the interface. Specifically,  in the crystal plasticity model, the slip resistance corresponding to the newly activated order parameter during the interface migration process needs to be reset, according to the statement in the zhao's article.

Please forgive me for using such words to express my views. If I have not expressed clearly, your questions are very welcome.

Wei

**Reply by PengWei97** [reply]
# Problem 2

> The grain average makes sense to me if you want to couple FFT and FEM.

I don't quite understand what you mean. 

The simulation of phase field and tensor mechanics field both use FEM to solve, which is adopted by me. According to my understanding, if different numerical solution methods are used, the complexity of coupling modeling is artificially increased, because the results obtained by different numerical solutions will have certain differences in data representation and data transmission also needs to be converted. 

At the same time, such a problem will not occur in FEM. It is only necessary to consider the mismatch of the two simulation methods for the description of the microstructure.

Thank you so much for the speedy reply.
Wei

**Reply by hugary1995** [reply]
I still believe equation 15 is computed per quadrature point...... I'm not convinced by sentences marked in green that they averaging over a grain, i.e. over multiple elements. Perhaps @dschwen @laagesen or @jiangwen84 who are more familiar with this work can chime in here.

> According to my understanding, the stress and internal parameters calculated by the tensor mechanics module are uniquely determined at each orthogonal point, without distinguishing which crystal grain they belong to.

No, only displacements and strain are unique, you can have arbitrarily many stresses at each quadrature point. In the phase field and the tensor mechanics modules, you can distinguish them using base names.

> However, I did not see how this method can reset the state parameters at the interface. Specifically, in the crystal plasticity model, the slip resistance corresponding to the newly activated order parameter during the interface migration process needs to be reset, according to the statement in the zhao's article.

Yeah, that's a fair question. If you could find equations describing how they are performing such reset, I'm happy to help!

**Reply by PengWei97** [reply]
> I still believe equation 15 is computed per quadrature point.

I think your understanding should be correct, because if you consider the average of multiple elements or multiple quadrature points, this will cause a stronger non-convergence factor, resulting in slow calculations or even direct non-convergence.

> you can have arbitrarily many stresses at each quadrature point.

According to your knowledge, in the `material` block, different stress values can exist at an quadrature point (especially at the boundary) at the same time, and these stress values belong to different `base_name`. But in the calculation and solution of `kernel\TensorMechanics`, only one value is needed for each orthogonal point. In Pattern1.i, it is realized by using `MultiPhaseStressMaterial`, which is achieved through an interpolation function on the phase field order parameters and weighted average.

Best Wishes
Wei

**Reply by PengWei97** [reply]
Yesterday I was trying to use the same method to consider the combination of linear elastic constitutive and phase field. The two modules can be calculated independently, but once they are combined, an error occurs. The error is as follows:
```bash
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property'g', requested by'elastic_free_energy' is not defined on block 0
Material property'g', requested by'elastic_free_energy_face' is not defined on block 0
Material property'g', requested by'elastic_free_energy_neighbor' is not defined on block 0
```

And I think the important code has been placed below:
```bash
[Variables]
  # Order parameter for the Matrix
  [./gr0]
    order = FIRST
    family = LAGRANGE
  [../]
  # Order parameters for the 2 different inclusion orientations
  [./gr1]
    order = FIRST
    family = LAGRANGE
  [../]

  # Mesh displacement
  [./disp_x]
    order = FIRST
    family = LAGRANGE
  [../]
  [./disp_y]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[Kernels]
  # Set up stress divergence kernels
  [./TensorMechanics]
  [../]
  [./PolycrystalKernel]
  [../]
  [./ACBulk_ElasticEnergy_gr0]
    type = AllenCahn
    variable = gr0
    args = 'gr1'
    mob_name = L
    f_name = Fe1
  [../]
  [./ACBulk_ElasticEnergy_gr1]
    type = AllenCahn
    variable = gr1
    args = 'gr0'
    mob_name = L
    f_name = Fe2
  [../]
[]

[Materials]
  [./Copper]
    type = GBEvolution
    block = 0
    T = 500 # K
    wGB = ${my_wGB} #15 # nm
    GBmob0 = 2.5e-6 #m^4/(Js) from Schoenfelder 1997
    Q = 0.23 # Migration energy in eV
    GBenergy = 0.708 # GB energy in J/m^2
    time_scale = ${my_time_scale}
    length_scale = ${my_length_scale}
    # 提供材料参数
  [../]

  # gr0
  [./elasticity_tensor_1]
    type = ComputeElasticityTensor
    base_name = phase1
    C_ijkl = '1.684e5 1.214e5 1.214e5 1.684e5 1.214e5 1.684e5 0.75e5 0.75e5 0.75e5'
    fill_method = symmetric9
    euler_angle_1 = 0.0
    euler_angle_2 = 0.0
    euler_angle_3 = 0.0
  [../]
  [./strain_1]
    type = ComputeSmallStrain
    base_name = phase1
    displacements = 'disp_x disp_y'
  [../]
  [./stress_1]
    type = ComputeLinearElasticStress
    base_name = phase1
  [../]

  # oversized phase
  [./elasticity_tensor_2]
    type = ComputeElasticityTensor
    base_name = phase2
    C_ijkl = '1.684e5 1.214e5 1.214e5 1.684e5 1.214e5 1.684e5 0.75e5 0.75e5 0.75e5'
    fill_method = symmetric9
    euler_angle_1 = 45.0
    euler_angle_2 = 0.0
    euler_angle_3 = 0.0
  [../]
  [./strain_2]
    type = ComputeSmallStrain
    base_name = phase2
    displacements = 'disp_x disp_y'
  [../]
  [./stress_2]
    type = ComputeLinearElasticStress
    base_name = phase2
  [../]

  # switching functions
  [./switching1]
    type = SwitchingFunctionMaterial
    function_name = h1
    eta = gr0
    h_order = SIMPLE
  [../]
  [./switching2]
    type = SwitchingFunctionMaterial
    function_name = h2
    eta = gr1
    h_order = SIMPLE
  [../]


  # #   # chemical free energies
  # [./chemical_free_energy_1]
  #   type = DerivativeParsedMaterial
  #   f_name = Fc1
  #   material_property_names = 'mu gamma_asymm'
  #   function = 'mu*( gr0^4/4.0 - gr0^2/2.0 + gr1^4/4.0 - gr1^2/2.0 + gamma_asymm*gr0^2*gr1^2) + 1.0/4.0'
  #   args = 'gr0 gr1' # use vector coupling
  #   derivative_order = 2
  #   enable_jit = true
  # [../]
  # [./chemical_free_energy_2]
  #   type = DerivativeParsedMaterial
  #   f_name = Fc2
  #   function = '1/4*gr1^4-1/2*gr1^2+gamma_asymm*gr0^2*gr1^2'
  #   args = 'gr1' # use vector coupling
  #   derivative_order = 2
  # [../]

  # [./barrier]
  #   type = MultiBarrierFunctionMaterial
  #   etas = 'eta1 eta2 eta3'
  # [../]

  # elastic free energies
  [./elastic_free_energy_1]
    type = ElasticEnergyMaterial
    base_name = phase1
    f_name = Fe1
    derivative_order = 2
    args = 'gr0' # should be empty
  [../]
  [./elastic_free_energy_2]
    type = ElasticEnergyMaterial
    base_name = phase2
    f_name = Fe2
    derivative_order = 2
    args = 'gr1' # should be empty
  [../]

  # global free energy
  [./elastic_free_energy]
    type = DerivativeMultiPhaseMaterial
    f_name = F # base name of the free energy function (used to name the material properties)
    fi_names = 'Fe1  Fe1' # List of free energies for the n phases
    hi_names = 'h1  h2' # Switching Function Materials that provide h(eta_i)
    etas     = 'gr0 gr1' # Order parameters for all phases.
    # args = 'c' # Arguments of the fi free energies - use vector coupling
    W = 0 # Energy barrier for the phase transformation from A to B
  [../]

  # Generate the global stress from the phase stresses
  [./global_stress]
    type = MultiPhaseStressMaterial
    phase_base = 'phase1 phase2'
    h          = 'h1     h2'
  [../]
[]
```
